File: 2017/08/rs/src/main.rs
use std::collections::HashMap;
use regex::Regex;

type Instruction = (String, String, Direction, i32, Operator, i32);

enum Direction {
    Inc,
    Dec,
}

impl Direction {
    fn parse(value: &str) -> Direction {
        match value {
            "inc" => Direction::Inc,
            "dec" => Direction::Dec,
            _ => panic!("Unknown direction '{}'", value),
        }
    }
}

enum Operator {
    Equal,
    NotEqual,
    LessThan,
    GreaterThan,
    LessOrEqual,
    GreatherOrEqual,
}

impl Operator {
    fn parse(value: &str) -> Operator {
        match value {
            "==" => Operator::Equal,
            "!=" => Operator::NotEqual,
            "<" => Operator::LessThan,
            ">" => Operator::GreaterThan,
            "<=" => Operator::LessOrEqual,
            ">=" => Operator::GreatherOrEqual,
            _ => panic!("Uknown operator '{}'", value),
        }
    }
}

fn is_condition_valid(source: i32, operator: &Operator, value: i32) -> bool {
    match operator {
        Operator::Equal => source == value,
        Operator::NotEqual => source != value,
        Operator::LessThan => source < value,
        Operator::GreaterThan => source > value,
        Operator::LessOrEqual => source <= value,
        Operator::GreatherOrEqual => source >= value,
    }
}

fn solve(instructions: &Vec<Instruction>) -> (i32, i32) {
    let mut memory = HashMap::new();
    let mut max_value = 0;
    for (target, source, direction, amount, operator, value) in instructions {
        let source_value = *memory.entry(source).or_insert(0);
        if !is_condition_valid(source_value, operator, *value) {
            continue;
        }
        *memory.entry(target).or_insert(0) += amount * match direction {
            Direction::Inc => 1,
            Direction::Dec => -1,
        };
        max_value = max_value.max(*memory.get(target).unwrap());
    };
    (*memory.values().max().unwrap(), max_value)
}

fn get_input(file_path: &String) -> Vec<Instruction> {
    let re = Regex::new(r"^(?P<target>[a-z]+)\s(?P<direction>inc|dec)\s(?P<amount>-?\d+)\sif\s(?P<source>[a-z]+)\s(?P<operator>==|!=|>=|<=|>|<)\s(?P<value>-?\d+)$").unwrap();
    std::fs::read_to_string(file_path).expect("Error reading input file!")
    .lines()
    .map(|line| re.captures(line).map(|cap| (
        String::from(cap.name("target").unwrap().as_str()),
        String::from(cap.name("source").unwrap().as_str()),
        Direction::parse(cap.name("direction").unwrap().as_str()),
        cap.name("amount").unwrap().as_str().parse().unwrap(),
        Operator::parse(cap.name("operator").unwrap().as_str()),
        cap.name("value").unwrap().as_str().parse().unwrap()
    )).unwrap())
    .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2017/08/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef enum {
    INCREMENT,
    DECREMENT
} Direction;
typedef enum {
    EQUAL,
    NOT_EQUAL,
    LESS_THAN,
    GREATER_THAN,
    LESS_OR_EQUAL,
    GREATER_OR_EQUAL
} Comparator;
typedef struct {
    int target, source, amount, value;
    Direction direction;
    Comparator comparator;
} Instruction;

typedef struct {
    int instructionCount, instructionCapacity;
    Instruction *instructions;
    int registerCount, registerCapacity;
    char **names;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int isConditionValid(int memoryValue, Comparator comparator, int value)
{
    switch (comparator)
    {
        case EQUAL:
            return memoryValue == value;
        case NOT_EQUAL:
            return memoryValue != value;
        case LESS_THAN:
            return memoryValue < value;
        case GREATER_THAN:
            return memoryValue > value;
        case LESS_OR_EQUAL:
            return memoryValue <= value;
        case GREATER_OR_EQUAL:
            return memoryValue >= value;
        default:
            perror("Unknow comparator to compare");
            exit(1);
    }
}

Results solve(Input input)
{
    int *memory = calloc(input.registerCount, sizeof(int));
    Instruction *instruction;
    int maxValue = 0;
    while (input.instructionCount--)
    {
        instruction = input.instructions++;
        if (isConditionValid(memory[instruction->source], instruction->comparator, instruction->value))
        {
            memory[instruction->target] += instruction->amount * (instruction->direction == INCREMENT ? 1 : -1);
            maxValue = memory[instruction->target] > maxValue ? memory[instruction->target] : maxValue;
        }
    }
    int currentMax = 0;
    for (int index = 0; index < input.registerCount; index++)
        currentMax = memory[index] > currentMax ? memory[index] : currentMax;
    free(memory);
    return (Results){currentMax, maxValue};
}

#define INPUT_INCREMENT 10

int getRegister(Input *input, const char *name)
{
    for (int index = 0; index < input->registerCount; index++)
        if (!strcmp(input->names[index], name))
            return index;
    if (input->registerCount == input->registerCapacity)
    {
        input->registerCapacity += INPUT_INCREMENT;
        char **oldNames = input->names;
        char **newNames = realloc(oldNames, input->registerCapacity * sizeof(char*));
        input->names = newNames;
    }
    input->names[input->registerCount] = malloc(strlen(name) + 1);
    strcpy(input->names[input->registerCount], name);
    return input->registerCount++;
}

Direction getDirection(const char *direction)
{
    if (!strcmp(direction, "inc"))
        return INCREMENT;
    if (!strcmp(direction, "dec"))
        return DECREMENT;
    perror("Unknow direction");
    perror(direction);
    exit(1);
}

Comparator getComparator(const char *comparator)
{
    if (!strcmp(comparator, "=="))
        return EQUAL;
    if (!strcmp(comparator, "!="))
        return NOT_EQUAL;
    if (!strcmp(comparator, "<"))
        return LESS_THAN;
    if (!strcmp(comparator, ">"))
        return GREATER_THAN;
    if (!strcmp(comparator, "<="))
        return LESS_OR_EQUAL;
    if (!strcmp(comparator, ">="))
        return GREATER_OR_EQUAL;
    perror("Unknown comparator");
    perror(comparator);
    exit(1);
}

void addToInput(Input *input, const char *target, const char *source, const char *comparator, const char *direction, int amount, int value)
{
    if (input->instructionCount == input->instructionCapacity)
    {
        input->instructionCapacity += INPUT_INCREMENT;
        Instruction *oldInstructions = input->instructions;
        Instruction *newInstructions = realloc(oldInstructions, input->instructionCapacity * sizeof(Instruction));
        input->instructions = newInstructions;
    }
    input->instructions[input->instructionCount] = (Instruction) {
        getRegister(input, target),
        getRegister(input, source),
        amount, value,
        getDirection(direction),
        getComparator(comparator)
    };
    input->instructionCount++;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    char *line = NULL;
    size_t lineLength;
    char target[5], direction[5], source[5], comparator[5];
    int amount, value;
    Input input = {
        0, INPUT_INCREMENT,
        malloc(INPUT_INCREMENT * sizeof(Instruction)),
        0, INPUT_INCREMENT,
        malloc(INPUT_INCREMENT * sizeof(char*))
    };
    while (getline(&line, &lineLength, file) != EOF)
    {
        if (sscanf(line, "%[^ ] %[^ ] %d if %[^ ] %[^ ] %d", &target, &direction, &amount, &source, &comparator, &value) == 6)
            addToInput(&input, target, source, comparator, direction, amount, value);
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    for (int index = 0; index < input.registerCount; index++)
        free(input.names[index]);
    free(input.names);
    free(input.instructions);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
